lib/charms/loki_k8s/v0/charm_logging.py

```diff
@@ -0,0 +1,327 @@
+    logging_endpoints = tuple(logging_endpoints)
+    sanitized_logging_endponts = []
+    for endpoint in logging_endpoints:
+        if isinstance(endpoint, str):
+            sanitized_logging_endponts.append(endpoint)
+        else:
+            errors.append(f"invalid endpoint: expected string, got {endpoint!r}")
+
+    if errors:
+        logger.error(
+            f"{charm}.{logging_endpoints_getter} should return an iterable of Loki push-api "
+            "(-compatible) endpoints (strings); "
+            f"ERRORS: {errors}"
+        )
+
+    return sanitized_logging_endponts
+
+
+def _get_server_cert(server_cert_getter, self, charm):
+    if isinstance(server_cert_getter, property):
+        server_cert = server_cert_getter.__get__(self)
+    else:  # method or callable
+        server_cert = server_cert_getter(self)
+
+    if server_cert is None:
+        logger.warning(
+            f"{charm}.{server_cert_getter} returned None; sending logs over INSECURE connection."
+        )
+        return None
+
+    if not Path(server_cert).is_absolute():
+        raise ValueError(
+            f"{charm}.{server_cert_getter} should return a valid tls cert absolute path (string | Path)); "
+            f"got {server_cert} instead."
+        )
+
+    if not Path(server_cert).exists():
+        logger.warning(f"cert not found at {server_cert}: sending logs over INSECURE connection")
+        return None
+
+    return server_cert
+
+
+def _setup_root_logger_initializer(
+    charm: Type[CharmBase],
+    logging_endpoints_getter: _GetterType,
+    server_cert_getter: Optional[_GetterType],
+    service_name: Optional[str] = None,
+):
+    """Patch the charm's initializer and inject a call to set up root logging."""
+    original_init = charm.__init__
+
+    @functools.wraps(original_init)
+    def wrap_init(self: CharmBase, framework: Framework, *args, **kwargs):
+        original_init(self, framework, *args, **kwargs)
+
+        if not is_enabled():
+            logger.info("Charm logging DISABLED by env: skipping root logger initialization")
+            return
+
+        try:
+            logging_endpoints = _get_logging_endpoints(logging_endpoints_getter, self, charm)
+        except Exception:
+            # if anything goes wrong with retrieving the endpoint, we go on with logging disabled.
+            # better than breaking the charm.
+            logger.exception(
+                f"exception retrieving the logging "
+                f"endpoint from {charm}.{logging_endpoints_getter}; "
+                f"proceeding with charm_logging DISABLED. "
+            )
+            return
+
+        if not logging_endpoints:
+            return
+
+        juju_topology = JujuTopology.from_charm(self)
+        tags = {
+            **juju_topology.as_dict(),
+            "service_name": service_name or self.app.name,
+            "charm_type_name": type(self).__name__,
+            "dispatch_path": os.getenv("JUJU_DISPATCH_PATH", ""),
```

Comment:
1) done.

2) `service_name` is a standard label that's picked up by grafana and has [a special meaning](https://community.grafana.com/t/how-to-get-value-of-service-name/113534), so we can't change that.
charm_type_name we could.

3) +1 for hook_name instead of dispatch_path